bdio-protobuf/src/main/java/com/blackducksoftware/bdio/proto/ProtobufBdioWriter.java

```diff
@@ -0,0 +1,154 @@
+			this.bdioArchive = new ZipOutputStream(outputStream);
+		}
+	}
+
+	/**
+	 * Write header entry to bdio archive
+	 *
+	 * @param protoHeader header to write
+	 * @throws IOException
+	 */
+	public void writeHeader(ProtoScanHeader protoHeader) throws IOException {
+		if (headerWritten) {
+			throw new IllegalStateException("BDIO header file has already been written to this archive");
+		}
+
+		bdioArchive.putNextEntry(new ZipEntry(HEADER_FILE_NAME));
+		bdioArchive.write(Shorts.toByteArray((short) BdioEntryType.HEADER.ordinal())); // bdio entry type
+		bdioArchive.write(Shorts.toByteArray(FORMAT_VERSION_2)); // format version
+
+		Any any = Any.pack(protoHeader);
+		any.writeTo(bdioArchive);
+
+		headerWritten = true;
+	}
+
+	/**
+	 * Write single file node for signature scan to bdio archive entry
+	 *
+	 * @param protoNode node to write
+	 * @throws IOException
+	 */
+	public void writeSignatureScanFileNode(ProtoFileNode protoNode) throws IOException {
+		// Ensure we're not surpassing the chunk size limit, adding 4 bytes to account
+		// for the length header
+		// for each message. This will not be 100% accurate since writeDelimitedTo uses
+		// varints for the header,
+		// which can be variable in size. See:
+		// https://developers.google.com/protocol-buffers/docs/encoding#varints
+		if ((bytesRemaining -= protoNode.getSerializedSize() + 4) > 0L) {
+			protoNode.writeDelimitedTo(bdioArchive);
+		} else {
+			bytesRemaining = MAX_CHUNK_SIZE;
+			bdioArchive.putNextEntry(new ZipEntry(String.format(ENTRY_FILE_NAME_TEMPLATE, entryCount++)));
+			bdioArchive.write(Shorts.toByteArray((short) BdioEntryType.CHUNK.ordinal())); // message type
+			bdioArchive.write(Shorts.toByteArray(FORMAT_VERSION_1)); // format version
+
+			if ((bytesRemaining -= protoNode.getSerializedSize() + 4) > 0L) {
+				protoNode.writeDelimitedTo(bdioArchive);
+			} else {
+				throw new RuntimeException("File metadata larger than maximum allowed chunk size");
+			}
+		}
+	}
+
+	/**
+	 * Write collection of nodes to bdio archive entry .
+	 *
+	 * @param protoNodes
+	 * @throws IOException
+	 */
+	public void writeBdioNodes(Collection<GeneratedMessageV3> protoNodes) throws IOException {
+		for (com.google.protobuf.GeneratedMessageV3 node : protoNodes) {
+			writeBdioNode(node);
+		}
+	}
+
+	/**
+	 * Write single node for to bdio archive entry .
+	 *
+	 * @param protoNode
+	 * @throws IOException
+	 */
+	public void writeBdioNode(com.google.protobuf.GeneratedMessageV3 protoNode) throws IOException {
+
+		Any any = Any.pack(protoNode);
+
+		if ((bytesRemaining -= protoNode.getSerializedSize() + 4) > 0L) {
+			any.writeDelimitedTo(bdioArchive);
+		} else {
+			bytesRemaining = MAX_CHUNK_SIZE;
+			bdioArchive.putNextEntry(new ZipEntry(String.format(ENTRY_FILE_NAME_TEMPLATE, entryCount++)));
+			bdioArchive.write(Shorts.toByteArray((short) BdioEntryType.CHUNK.ordinal())); // bdio entry type
+			bdioArchive.write(Shorts.toByteArray(FORMAT_VERSION_2)); // format version
```

Comment:
Similar comment as the reader.   Having versioned writers would be similar in advantage to versioned readers.

Reply:
Is my understanding correct that we are going to add writeBdioNodeV1, etc (for each version except last) and writeBdioNode for latest version? Why do we need to provide writers for previous versions?
So, far I moved the common functionality that can be shared among different writeBdioNode methods to private method, since I don't see in the future any changes in how we write data (the only thing that is going to change is version number and validation, which is handled by separate validator per version).